File: backend/src/main/kotlin/niconicotagger/dto/api/response/SongsWithPvsResponse.kt
package niconicotagger.dto.api.response

import niconicotagger.dto.api.misc.VocaDbSongEntryWithPvs
import niconicotagger.dto.inner.misc.SongType

data class SongsWithPvsResponse(
    override val items: List<VocaDbSongEntryWithPvs>,
    val songTypeStats: Map<SongType, Int>,
    override val totalCount: Long,
) : ApiResponse<VocaDbSongEntryWithPvs>

File: frontend/src/main/typescript/backend/dto/response/songsWithPvsResponse.ts
import { SongTypeStats } from "@/backend/dto/songTypeStats";
import { NndTagData, UpdateErrorReport, VocaDbTagSelectable } from "@/backend/dto/lowerLevelStruct";
import { SongType } from "@/backend/dto/enumeration";

export interface SongsWithPvsResponse {
  items: VocaDbSongEntryWithPvs[];
  songTypeStats: SongTypeStats;
  totalCount: number;
}

export interface VocaDbSongEntryWithPvs {
  entry: SongEntry;
  availablePvs: PvWithSuggestedTags[];
  unavailablePvs: UnavailableNndVideo[];
  toUpdate: boolean;
  visible: boolean;
  tagIdsToAssign: number[];
  errorReport: UpdateErrorReport | null;
}

export interface SongEntry {
  id: number;
  name: string;
  type: string;
  artistString: string;
  publishedOn: Date;
}

export interface PvWithSuggestedTags {
  video: AvailableNndVideo;
  suggestedTags: VocaDbTagSelectable[];
  visible: boolean;
}

export interface UnavailableNndVideo {
  id: string;
  title: string;
  error: string;
  toDisable: boolean;
}

export interface AvailableNndVideo {
  id: string;
  title: string;
  description: string | null;
  tags: NndTagData[];
}
